stepsfunction/handler.py

```diff
@@ -163,19 +178,186 @@ def ocr_page(event, context):
     bucket = s3rec['bucket']['name']
     key = s3rec['object']['key']
     _doc_pdf, jid, name_pdf = key.split('/')
     LOG.info(f'bucket={bucket} key={key} jid={jid} name_pdf={name_pdf}')
 
-    worker_name = "notify_upload_worker"
-    sf = boto3.client('stepfunctions')
+    detected_text = get_textract_data(bucket, key)
+    write_textract_to_s3(detected_text, bucket, key)
+    pdf_upload = PDFUpload.get(hash_key=jid)
+    update_db_after_ocr(pdf_upload, detected_text, name_pdf, jid)
+
     # Pretend we've discovered we've finished all the pages
-    if name_pdf == '0000.pdf':
-        res = sf.get_activity_task(activityArn=activity_arn, workerName=worker_name)
-        LOG.info(f'sf_activity={res}')
-        sf_input = dumps(res['input'])
-        res = sf.send_task_success(taskToken=res['taskToken'], output=sf_input)
+    if check_orc_done(pdf_upload, jid):
+        # update db
+        pdf_upload.update_with_log(
+            actions=[
+                PDFUpload.status.set("ocred"),
+                PDFUpload.updatedAt.set(datetime.now())
+            ])
+        send_task_ocr_activity(jid)
+
+
+def get_textract_data(bucket, key):
+    """Using AWS textract."""
+    LOG.info(f'Loading get_textract_data bucket:{bucket}, key:{key}')
+    # convert to jpg
+    jpg_file = extract_jpg_from_pdf(bucket, key)
+    # Call Amazon Textract
+    tx = boto3.client(
+        service_name='textract',
+        region_name='us-east-1',
+        endpoint_url='https://textract.us-east-1.amazonaws.com',
+    )
+    response = tx.detect_document_text(
+        Document={
+            'Bytes': jpg_file
+        })
+    detected_text = ''
+
+    # Print detected text
+    for item in response['Blocks']:
+        if item['BlockType'] == 'LINE':
+            detected_text += item['Text'] + '\n'
+    return detected_text
+
+
+def write_textract_to_s3(textract_data, bucket, key):
+    """Save detected text to S3."""
+    LOG.info(f'Loading write_textract_to_s3 bucket:{bucket}, key:{key}')
+    generate_path = os.path.splitext(key)[0] + '.txt'
+    s3client = boto3.client('s3')
+    s3client.put_object(Body=textract_data, Bucket=bucket, Key=generate_path)
+    LOG.info(f'generateFilePath: {generate_path}')
+
+
+def update_db_after_ocr(pdf_upload, content, name, jid):
+    """Update PDFUpload after finishing orc."""
+    page = SinglePage(
+        page_id=name[:4],
+        content=content
+    )
+    pages = pdf_upload.pages
+    if pages:
+        pages.append(page)
+    else:
+        pages = [page]
+    LOG.info(f'loading update_db_after_ocr pages:{pages}')
+    pdf_upload.update_with_log(
+        actions=[
+            PDFUpload.pages.set(pages),
+            PDFUpload.updatedAt.set(datetime.now())
+        ])
+
+
+def check_orc_done(pdf_upload, jid):
+    """Get db & check done."""
+    LOG.info(f'loading check_orc_done')
+    num_pages = pdf_upload.num_pages
+    pages = pdf_upload.pages
```

Comment:
This will work for now, but it's an expensive way to do it. Imagine if our doc has 3000 pages. And each time a new page is added, it makes a DB read to get all the (up to) 3000 page objects.  

It is also possible to have a "race condition" here: if there are many Lambdas all doing this at once, one may read old data and not get new data that a different lambda has just written.

It took me a while to come up with a efficient and race-condition-free way of doing this, and it uses native DDB features which your model abstraction doesn't support.  I really need to get time to talk to you about it and walk you through the code. 

Again, this is fine for now, we can replace the model and detection of "done" later on. First: get it running; Next: get it running reliably. Finally: get it running fast. :-)